overlord/snapstate/snapstate.go

```diff
@@ -2306,7 +2307,44 @@ func canRemove(st *state.State, si *snap.Info, snapst *SnapState, removeAll bool
 		rev = si.Revision
 	}
 
-	return PolicyFor(si.Type(), deviceCtx.Model()).CanRemove(st, snapst, rev, deviceCtx)
+	if err := PolicyFor(si.Type(), deviceCtx.Model()).CanRemove(st, snapst, rev, deviceCtx); err != nil {
+		return err
+	}
+
+	// check if this snap is required by any validation set in enforcing mode
+	enforcedSets, err := EnforcedValidationSets(st)
+	if err != nil {
+		return err
+	}
+	if enforcedSets == nil {
+		return nil
+	}
+	requiredValsets, requiredRevision, err := enforcedSets.CheckPresenceRequired(si)
+	if err != nil {
+		if _, ok := err.(*snapasserts.PresenceConstraintError); !ok {
+			return err
+		}
+		// else - presence is invalid, nothing to do (not really possible since
+		// it shouldn't be allowed to get installed in the first place).
+		return nil
+	}
+	if len(requiredValsets) == 0 {
+		// not required by any validation set (or is optional)
+		return nil
+	}
+	// removeAll is set if we're removing the snap completely
+	if removeAll {
+		if requiredRevision.Unset() {
+			return fmt.Errorf("snap %q is required by validation sets: %s", si.InstanceName(), strings.Join(requiredValsets, ","))
+		}
+		return fmt.Errorf("snap %q at revision %s is required by validation sets: %s", si.InstanceName(), requiredRevision, strings.Join(requiredValsets, ","))
+	}
+
+	// rev is set at this point (otherwise we would hit removeAll case)
+	if requiredRevision.N == rev.N {
+		return fmt.Errorf("snap %q at revision %s is required by validation sets: %s", si.InstanceName(), rev, strings.Join(requiredValsets, ","))
```

Comment:
this case doesn't seem to be tested

Reply:
Indeed, thanks, added with a question (see XXX) whether this is a real scenario.